### PR TITLE
Added citation types

### PR DESCRIPTION
**Pull request by il8677**
This change revamps the code to allow for easy addition of different citation types in an encapsulated manner (You add one definition to an array, and it works out of the box). It moves handling from the submit button handler to the class itself, allowing for values being used in the citation to be dynamically defined on a instance-to-instance basis.
This looks like a big scary change and it is, all of the core functionality has been modified:

_Note: This makes the serialisation stop working, however, it should not be too difficult to get the running again._

### Javascript
**Web Citation**
Previously, this class was very rigid, it was statically constructed by a defined set of member variables, and any code that interfaced with it was accessing member variables directly. This system did not allow encapsulated expansion of different web citation types with different sets of data required: Only one set of variables were able to be given. Several changes have been made to accomodate for this feature and to allow future expansion:

- Class constructor now takes **only** the index of the citation format
- Class constructor takes the values of the elements specified in citationType (more on this later) instance from the form directly into an array (based on HTML id) called "values"
    - The values array is dynamically populated: the contents are not constrained by type of input. APA might have date at index 0, when MLA might have the name at index 0. This allows arbitrary definition of what data is needed as outlined in the "citationType" class
    - In essence, the value array contains all the user input required based on the currently selected citation format

- citationHTML() now looks at a citation format string (defined in "citationType" class), then returns the value of that.
    - To allow choice of format of citation, in a simple define-once and every thing works way. This function parses the citationType.citationFormat string for special escape characters that define where data from the values array should be inserted

**Citation Format**
New class that contains definitions of citation type:

- Constructed with 3 (+1 optional) arguments
    - Name: The name of the citation format
    - enabledElements: Array of HTML IDs (string) of the elements in the form that the user is allowed to input into and values to be taken and insert into the values array.
    - Format: A string containing the format for the citation. Everything that isn't escaped by a "$" character is put in. Where the character after the "$" is _n_, the citationHTML function takes "$" as "Insert string at values[n]". As this is directly inserted into html, HTML tags work.
    - allowedElements = []: This is an optional argument specifies the id of which form elements to allow the user to _input_ into. The difference between this and enabledElements is that the latter _will_ insert the form value into the values array, while elements specified in this _will not_ be inserted in the values array. This is useful for data that is still needed, but is not needed in the format string. Example of this later on.

**Global variables**
The global variables added:

- citationFormats: Array that contains CitationFormats instances
- citationIndex: The index of citationFormats that the user currently has selected

**setCitationType(i)**
This function is called when the user chooses a citation format type from the drop-down menu.

- Takes 1 argument:
    - i: The index of the new selected citation
- Disable ALL form elements (that have the class ".form_control", which is bootstrap for form elements, so all of them).
- Re-enable the elements as specified in the enabledElements array of the new selected citation
- Re-enable the elements specified in the allowedElements array of the new selected citation
- Sets the citationIndex variable to i

**Startup function**
Very simple, iterates over each CitationFormat instance in citationFormats array, then wrapping the name in <button> tags that call the setCitationType when called. Adding them to the drop down menu defined in the HTML file.

**Citation submit button handler**
This is what is called when the submit button is pressed. It was already existing. Now it has a few changes.

- It no longer pushes the citation array with a new citation with the values inside the array, instead it just provides the citationIndex
- It updates two hidden forms with the MLA format dates.
_Notes_
Using hidden forms and updating them after calculating from user input form data is very useful, and how you would have user input numerical values, but still perform calculation on them. For example, if you directly took the date used in MLA citation from the date field input, the format would not be MLA compliant (Numbers, not name of month). Instead, you specify the hidden forms in "enabledForms" for MLA, and the date forms in the "allowedForms" (so the user can still input into them), then the new citation will take the values you just set in the hidden forms which are set from the MLADate() function.

### HTML
Two very small changes:
- The order of which the libraries are loaded are changed, loading jquery, then a new library popper, and then finally bootstrap. This is justified and required because:
    - Popper is a bootstrap recommend library required for bootstraps dropdown menu
    - Jquery had to be loaded before popper
    - Bootstrap had to be after everything
- A standard bootstrap dropdown menu is added, this is empty by default but is populated by javascript function.

### CSS
This just has definitions for the buttons in the dropdown menu, so they don't look bad. They still look bad but they don't look terrible. Further work should be put to make sure they fit the style, this is mainly a client side backend pull request.

### CODE REVIEW

File: assets/js/main.js [hunk 1]
```diff
@@ -9,37 +9,73 @@ function htmlEscape(str) {
...
+        this.format = format;
+        this.enabledElements = enabledElements;
+        this.allowedElements = allowedElements;
+    }
+}
+//Dictionary:
+//authorFirst           authorLast
+//PageTitle             siteTitle           manualCiteURL
+//date[Published/Accessed][Day/Month/Year]
+//MLA[Publish/Access]Date -> Hidden form with processed dates
```
Comment (iDoObject) [inline assets/js/main.js]: What are you trying to say here?

File: assets/js/main.js [hunk 2]
```diff
@@ -9,37 +9,73 @@ function htmlEscape(str) {
...
+
+let citationFormats = [
+    new CitationFormat("MLA 7",
```
Comment (iDoObject) [inline assets/js/main.js]: Should be `"MLA7"` and `"MLA8"` (no spaces)

File: assets/js/main.js (hunk 1 above)
Comment (il8677) [reply assets/js/main.js]: List of id in the form, makes it easier to know what to add, instead of constantly changing files.

**Comment by iDoObject** [comment]
Have you considered a more OOP-ish approach:
```

Citation
├── BookCitation
└── WebCitation
    ├── APACitation
    ├── MLA7Citation
    └── MLA8Citation
```

**Comment by il8677** [comment]
This feature can be added in the future quite easily (without base classes, inheritance would not add anything). But this provides a base for adding different types.

**Pull request merged**